### PR TITLE
__beaconCallback(uint256)

### PR DESCRIPTION
**Pull request by pdyraga**
Closes: #1527 

See https://github.com/keep-network/keep-common/pull/22
See https://github.com/keep-network/keep-ecdsa/pull/339

Instead of allowing the customer to pass callback signature of their choice, we now expect that the callback contract delcares public `__beaconCallback(uint256)` function.

Previously, `KeepRandomBeaconServiceImplV1` allowed senders to specify an arbitrary method that would receive a callback once the beacon generated a relay entry. Arbitrary callbacks could be used to force the service contract to execute many functions within the keep contract system. For example, functions with `onlyServiceContract` modifier in the operator contract.

We still allow specifying an address of the callback contract. This could be beneficial in a situations where one contract pays for a random number for another contract.

### CODE REVIEW

**Comment by pdyraga** [comment]
Tested locally with a relay request without a callback and against `keep-ecdsa` as updated in https://github.com/keep-network/keep-ecdsa/pull/339; everything works as expected.

File: solidity/contracts/KeepRandomBeaconServiceImplV1.sol [hunk 1]
```diff
@@ -129,7 +128,7 @@ contract KeepRandomBeaconServiceImplV1 is DelayedWithdrawal, ReentrancyGuard, IR
         _pendingWithdrawal = 0;
         _previousEntry = _beaconSeed;
         _registry = registry;
-        _baseCallbackGas = 18789;
+        _baseCallbackGas = 10961;
```
Comment (ngrinkevich) [inline solidity/contracts/KeepRandomBeaconServiceImplV1.sol]: always nice to see discounts 😄☝️

File: solidity/contracts/examples/CallbackContract.sol [hunk 2]
```diff
@@ -9,19 +9,12 @@ contract CallbackContract {
 
     uint256 internal _lastEntry;
 
-    /**
-     * @dev Example of a callback method. Method signature can be
-     * calculated as bytes4(keccak256("callback(uint256)")
-    */
-    function callback(uint256 requestResponse)
+    function __beaconCallback(uint256 entry)
         public
```
Comment (dimpar) [inline solidity/contracts/examples/CallbackContract.sol]: do we want to have `public` in a new line?

Comment (pdyraga) [reply solidity/contracts/examples/CallbackContract.sol]: This is how our Solidity linters work in `keep-ecdsa`, see [here](https://github.com/keep-network/keep-ecdsa/blob/master/solidity/contracts/BondedECDSAKeepFactory.sol). I don't think we ported this work to `keep-core` already but we'll be doing it soon. And then we'll have less work here.

Comment (dimpar) [reply solidity/contracts/examples/CallbackContract.sol]: hah.. I see that it is kinda random in `keep-ecdsa`. I guess wip :) anyway, not biggy, we can leave it as is then

Comment (pdyraga) [reply solidity/contracts/examples/CallbackContract.sol]: > hah.. I see that it is kinda random in keep-ecdsa

Yep. We should all have pre-commit hooks installed with the same linting configuration but we want to make it enforced on the build level.

File: solidity/scripts/request-relay-entry-with-callback.js [hunk 3]
```diff
@@ -1,10 +1,9 @@
-const crypto = require("crypto")
 const KeepRandomBeaconServiceImplV1 = artifacts.require("KeepRandomBeaconServiceImplV1.sol");
 const KeepRandomBeaconService = artifacts.require('KeepRandomBeaconService.sol');
 
 // Example usage:
-// truffle exec ./scripts/request-relay-entry-with-callback.js yourContractAddress "callbackMethodName" callbackGas
-// truffle exec ./scripts/request-relay-entry-with-callback.js 0x9F57C01059057d821c6b4B04A4598322661C934F "callback(uint256)" 20000
+// truffle exec ./scripts/request-relay-entry-with-callback.js yourContractAddress callbackGas
+// truffle exec ./scripts/request-relay-entry-with-callback.js 0x9F57C01059057d821c6b4B04A4598322661C934F 20000
```
Comment (dimpar) [inline solidity/scripts/request-relay-entry-with-callback.js]: I think this won't work, the number of params has changed. I've changed the script to this:
```
let entryFeeEstimate = await contractInstance.entryFeeEstimate(process.argv[5]);
    let tx = await contractInstance.methods['requestRelayEntry(address,uint256)'](
      process.argv[4],
      process.argv[5],
      {value: entryFeeEstimate}
    )
```
and then the request went trough

Comment (pdyraga) [reply solidity/scripts/request-relay-entry-with-callback.js]: Good catch! Fixed in https://github.com/keep-network/keep-core/pull/1532/commits/767add64f9002e5e736ed13a9f5d2e957e908646

Comment (dimpar) [reply solidity/scripts/request-relay-entry-with-callback.js]: we also need to change:
`await contractInstance.entryFeeEstimate(process.argv[6])` to
`await contractInstance.entryFeeEstimate(process.argv[5])`

Comment (pdyraga) [reply solidity/scripts/request-relay-entry-with-callback.js]: 🤦‍♂ Done.

**Comment by dimpar** [comment]
Are we sure that  `20000` of `callbackGas` is enough? I tried it a couple of times and then called `CallbackContract::lastEntry()` and the number didn't change. Only when I added gas `0` at the end (`200000`), then a new entry was set in the callback contract.

**Comment by pdyraga** [comment]
> Are we sure that `20000` of `callbackGas` is enough? I tried it a couple of times and then called `CallbackContract::lastEntry()` and the number didn't change. Only when I added gas (`200000`), then a new entry was set in the callback contract. `200000` is probably too much, but just wanted to do a quick validation.

Depends on the callback. We don't hardcode `20000` anywhere in contracts. In `keep-ecdsa` we need [`41830`](https://github.com/keep-network/keep-ecdsa/blob/master/solidity/contracts/BondedECDSAKeepFactory.sol#L74) in the worst case. Each callback should be first checked with `estimateGas`.

**Comment by dimpar** [comment]
maybe it would be good to add in `@dev` in `CallbackCotract.sol` that it will require `30k`?

anyway, to sum up:
tested with our `CallbackCotract.sol` deployed. It required `30k` of callback gas. Relay entry was generated and set. Looks good!

**Pull request merged**